Core/src/org/sleuthkit/autopsy/communications/ContactCache.java

```diff
@@ -0,0 +1,160 @@
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import java.beans.PropertyChangeListener;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import org.sleuthkit.autopsy.casemodule.Case;
+import org.sleuthkit.autopsy.coreutils.Logger;
+import org.sleuthkit.autopsy.ingest.IngestManager;
+import static org.sleuthkit.autopsy.ingest.IngestManager.IngestModuleEvent.DATA_ADDED;
+import org.sleuthkit.autopsy.ingest.ModuleDataEvent;
+import org.sleuthkit.datamodel.Account;
+import org.sleuthkit.datamodel.BlackboardArtifact;
+import org.sleuthkit.datamodel.BlackboardAttribute;
+import org.sleuthkit.datamodel.TskCoreException;
+
+/**
+ * A singleton cache of the Contact artifacts for accounts. The map of account
+ * unique ids to list of contact artifacts is stored in a LoadingCache which
+ * expires after 10 of non-use. 
+ *
+ */
+final class ContactCache {
+
+    private static final Logger logger = Logger.getLogger(ContactCache.class.getName());
+
+    private static ContactCache instance;
+    
+    private final LoadingCache<String, Map<String, List<BlackboardArtifact>>> accountMap;
+
+    /**
+     * Returns the list of Contacts for the given Account.
+     *
+     * @param account Account instance.
+     *
+     * @return List of TSK_CONTACT artifacts that references the given Account.
+     *         An empty list is returned if no contacts are found.
+     *
+     * @throws ExecutionException
+     */
+    static synchronized List<BlackboardArtifact> getContacts(Account account) throws ExecutionException {
+        return getInstance().accountMap.get("realMap").get(account.getTypeSpecificID());
+    }
+
+    /**
+     * Force the cache to invalidate all entries.
+     */
+    static synchronized void invalidateCache() {
+        getInstance().accountMap.invalidateAll();
+    }
+
+    /**
+     * Construct a new instance.
+     */
+    private ContactCache() {
+        
+        accountMap = CacheBuilder.newBuilder().expireAfterAccess(10, TimeUnit.MINUTES).build(
+                new CacheLoader<String, Map<String, List<BlackboardArtifact>>>() {
+            @Override
+            public Map<String, List<BlackboardArtifact>> load(String key) {
+                try {
+                    return  buildMap();
+                } catch (SQLException | TskCoreException ex) {
+                    logger.log(Level.WARNING, "Failed to build account to contact map", ex);
+                } 
+                return new HashMap<>();  // Return an empty map if there is an exception to avoid NPE and continual trying.
+            }
+        });
+
+        PropertyChangeListener ingestListener = pce -> {
```

Comment:
In its current state, switching between cases may return old artifacts. Could you invalidate the cache when a case is closed?

Reply:
Good idea.